File: apps/android/app/src/main/java/com/galaxyssi/chat/AgentReplySpeechController.kt
package com.galaxyssi.chat

import com.galaxyssi.chat.voice.modelstream.CommittedSpeechChunk
import com.galaxyssi.chat.voice.modelstream.DefaultSentenceCommitter
import com.galaxyssi.chat.voice.modelstream.SpeechTextNormalizer

internal data class AgentReplySpeechTarget(
    val responseId: String,
    val entryId: String,
    val text: String,
    val complete: Boolean
)

internal data class AgentReplySpeechCommand(
    val cancelSessionId: String = "",
    val beginSessionId: String = "",
    val chunks: List<CommittedSpeechChunk> = emptyList(),
    val finishSessionId: String = "",
    val scheduleCommitSessionId: String = "",
    val changedEntryIds: Set<String> = emptySet()
)

internal object AgentReplySpeechPresentationPolicy {
    fun latestTarget(entries: List<AgentTranscriptEntry>): AgentReplySpeechTarget? = entries
        .asReversed()
        .asSequence()
        .filter { entry -> entry.role == AgentTranscriptRole.ASSISTANT }
        .mapNotNull(::target)
        .firstOrNull()

    fun target(entry: AgentTranscriptEntry): AgentReplySpeechTarget? {
        if (entry.role != AgentTranscriptRole.ASSISTANT ||
            entry.dedupeKey.startsWith("approval:") ||
            entry.dedupeKey.startsWith("agent-recovery:")
        ) return null
        val text = speakableText(entry)
        if (text.isBlank()) return null
        return AgentReplySpeechTarget(
            responseId = responseId(entry),
            entryId = entry.id,
            text = text,
            complete = !entry.id.startsWith("agent-stream-")
        )
    }

    fun speakableText(entry: AgentTranscriptEntry): String {
        val plain = CodexStyleResponsePolicy.sanitizeAssistantText(entry.text)
        if (plain.isNotBlank()) return plain
        return AgentRichContentCodec.decode(entry.richOutputJson)
            .mapNotNull { block ->
                when (block.type) {
                    AgentRichBlockType.TEXT,
                    AgentRichBlockType.HEADING,
                    AgentRichBlockType.QUOTE -> block.text.ifBlank { block.title }
                    AgentRichBlockType.LIST -> block.rows.joinToString("\n") { row ->
                        row.getOrNull(1).orEmpty()
                    }
                    else -> null
                }?.takeIf(String::isNotBlank)
            }
            .joinToString("\n\n")
    }

    private fun responseId(entry: AgentTranscriptEntry): String = sequenceOf(
        entry.turnId,
        entry.taskId,
        entry.dedupeKey,
        entry.id
    ).map(String::trim).first(String::isNotBlank)
}

internal class AgentReplySpeechController {
    private data class Session(
        var target: AgentReplySpeechTarget,
        var playbackSessionId: String = "",
        val committer: DefaultSentenceCommitter = DefaultSentenceCommitter(),
        var enabled: Boolean = false,
        var observedText: String = target.text,
        var deltaSequence: Long = 0L,
        var inputClosed: Boolean = false
    )

    private var active: Session? = null
    private var playbackSequence = 0L

    fun observe(target: AgentReplySpeechTarget?): AgentReplySpeechCommand {
        val previous = active
        if (target == null) {
            active = null
            return AgentReplySpeechCommand(
                cancelSessionId = previous?.takeIf(Session::enabled)?.playbackSessionId.orEmpty(),
                changedEntryIds = setOfNotNull(previous?.target?.entryId)
            )
        }
        if (previous == null || previous.target.responseId != target.responseId) {
            active = Session(target)
            return AgentReplySpeechCommand(
                cancelSessionId = previous?.takeIf(Session::enabled)?.playbackSessionId.orEmpty(),
                changedEntryIds = setOfNotNull(previous?.target?.entryId, target.entryId)
            )
        }

        val oldEntryId = previous.target.entryId
        val wasComplete = previous.target.complete
        val delta = appendedText(previous.observedText, target.text)
        previous.target = target
        previous.observedText = target.text
        val chunks = if (previous.enabled && delta.isNotEmpty() && !previous.inputClosed) {
            previous.committer.acceptDelta(previous.deltaSequence++, delta)
        } else {
            emptyList()
        }.toMutableList()
        val shouldFinish = previous.enabled && target.complete && !wasComplete && !previous.inputClosed
        if (shouldFinish) {
            chunks += previous.committer.flush()
            previous.inputClosed = true
        }
        return AgentReplySpeechCommand(
            chunks = chunks,
            finishSessionId = previous.playbackSessionId.takeIf { shouldFinish }.orEmpty(),
            scheduleCommitSessionId = previous.playbackSessionId.takeIf {
                previous.enabled && !target.complete && delta.isNotEmpty()
            }.orEmpty(),
            changedEntryIds = setOfNotNull(
                oldEntryId.takeIf { it != target.entryId },
                target.entryId.takeIf { it != oldEntryId }
            )
        )
    }

    fun toggle(target: AgentReplySpeechTarget): AgentReplySpeechCommand {
        val session = sessionFor(target)
        if (session.enabled) {
            session.enabled = false
            session.inputClosed = false
            return AgentReplySpeechCommand(
                cancelSessionId = session.playbackSessionId,
                changedEntryIds = setOf(session.target.entryId)
            )
        }
        return begin(session, target.text, target.complete)
    }

    fun readFromParagraph(
        target: AgentReplySpeechTarget,
        paragraph: String,
        sourceText: String = target.text,
        startOffset: Int = sourceText.indexOf(paragraph)
    ): AgentReplySpeechCommand {
        val speech = SpeechTextNormalizer.normalize(
            continuationFromParagraph(target.text, paragraph, sourceText, startOffset)
        )
        if (speech.isBlank()) return AgentReplySpeechCommand()
        val session = sessionFor(target)
        session.observedText = target.text
        return begin(session, speech, target.complete)
    }

    fun stop(): AgentReplySpeechCommand {
        val session = active?.takeIf(Session::enabled) ?: return AgentReplySpeechCommand()
        session.enabled = false
        session.inputClosed = false
        return AgentReplySpeechCommand(
            cancelSessionId = session.playbackSessionId,
            changedEntryIds = setOf(session.target.entryId)
        )
    }

    fun isPlaying(): Boolean = active?.enabled == true

    fun commitDue(sessionId: String): AgentReplySpeechCommand {
        val session = active?.takeIf {
            it.playbackSessionId == sessionId && it.enabled && !it.inputClosed
        } ?: return AgentReplySpeechCommand()
        val chunks = session.committer.commitDue()
        return AgentReplySpeechCommand(chunks = chunks)
    }

    fun disable(sessionId: String): Set<String> {
        val session = active?.takeIf { it.playbackSessionId == sessionId } ?: return emptySet()
        session.enabled = false
        session.inputClosed = false
        return setOf(session.target.entryId)
    }

    fun isEnabled(target: AgentReplySpeechTarget): Boolean = active?.let { session ->
        session.target.responseId == target.responseId && session.enabled
    } == true

    fun isActive(target: AgentReplySpeechTarget): Boolean =
        active?.target?.responseId == target.responseId

    private fun sessionFor(target: AgentReplySpeechTarget): Session {
        val current = active
        if (current != null && current.target.responseId == target.responseId) {
            current.target = target
            current.observedText = target.text
            return current
        }
        return Session(target).also { active = it }
    }

    private fun begin(
        session: Session,
        initialText: String,
        complete: Boolean
    ): AgentReplySpeechCommand {
        val previousPlaybackSessionId = session.playbackSessionId.takeIf { session.enabled }.orEmpty()
        session.playbackSessionId = playbackSessionId(session.target.responseId, ++playbackSequence)
        session.enabled = true
        session.inputClosed = false
        session.deltaSequence = 0L
        session.committer.reset(session.playbackSessionId)
        val chunks = session.committer.acceptDelta(session.deltaSequence++, initialText).toMutableList()
        if (complete) {
            chunks += session.committer.flush()
            session.inputClosed = true
        }
        return AgentReplySpeechCommand(
            cancelSessionId = previousPlaybackSessionId,
            beginSessionId = session.playbackSessionId,
            chunks = chunks,
            finishSessionId = session.playbackSessionId.takeIf { complete }.orEmpty(),
            scheduleCommitSessionId = session.playbackSessionId.takeUnless { complete }.orEmpty(),
            changedEntryIds = setOf(session.target.entryId)
        )
    }

    private fun appendedText(previous: String, current: String): String = when {
        current == previous -> ""
        current.startsWith(previous) -> current.substring(previous.length)
        previous.startsWith(current) -> ""
        else -> ""
    }

    private fun continuationFromParagraph(
        targetText: String,
        paragraph: String,
        sourceText: String,
        requestedStartOffset: Int
    ): String {
        val startOffset = requestedStartOffset.takeIf { it in 0..sourceText.length }
        if (startOffset != null) {
            val sourceStart = if (sourceText == targetText) {
                0
            } else {
                targetText.indexOf(sourceText).takeIf { it >= 0 }
            }
            if (sourceStart != null) {
                return targetText.substring((sourceStart + startOffset).coerceAtMost(targetText.length))
            }
        }
        val paragraphStart = targetText.indexOf(paragraph)
        return if (paragraphStart >= 0) targetText.substring(paragraphStart) else paragraph
    }

    private fun playbackSessionId(responseId: String, sequence: Long): String =
        "agent-reply-${responseId.hashCode().toUInt().toString(16)}-$sequence"
}

File: apps/ios/GalaxySSI/AgentReplySpeechController.swift
import CryptoKit
import Foundation

struct AgentReplySpeechTarget: Equatable {
  var responseId: String
  var entryId: String
  var text: String
  var complete: Bool
}

struct AgentReplySpeechCommand: Equatable {
  var cancelSessionId = ""
  var beginSessionId = ""
  var appendedText = ""
  var finishSessionId = ""
  var scheduleCommitSessionId = ""
  var changedEntryIds: Set<String> = []
}

enum AgentReplySpeechPresentationPolicy {
  static func latestTarget(_ messages: [ChatMessage]) -> AgentReplySpeechTarget? {
    messages.reversed().compactMap(target).first
  }

  static func target(_ message: ChatMessage) -> AgentReplySpeechTarget? {
    guard !message.isMine,
          !message.isSystem,
          !excludedRemoteMessagePrefixes.contains(where: message.remoteMessageId.hasPrefix) else {
      return nil
    }
    let text = speakableText(message)
    guard !text.isBlank else { return nil }
    return AgentReplySpeechTarget(
      responseId: responseId(message),
      entryId: message.id.uuidString.lowercased(),
      text: text,
      complete: !message.remoteMessageId.hasPrefix("agent-stream-")
    )
  }

  static func speakableText(_ message: ChatMessage) -> String {
    let plain = CodexStyleResponsePolicy.sanitizeAssistantText(message.content)
      .trimmingCharacters(in: .whitespacesAndNewlines)
    if !plain.isEmpty { return plain }
    return AgentRichContentCodec.decode(message.richOutputJson)
      .compactMap { block -> String? in
        switch block.type {
        case .text, .heading, .quote:
          return block.text.ifBlank(block.title)
        case .list:
          return block.rows.map { $0.dropFirst().joined(separator: " ") }.joined(separator: "\n")
        default:
          return nil
        }
      }
      .filter { !$0.isBlank }
      .joined(separator: "\n\n")
  }

  private static func responseId(_ message: ChatMessage) -> String {
    [message.turnId, message.remoteMessageId, message.conversationId, message.id.uuidString]
      .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
      .first { !$0.isEmpty } ?? message.id.uuidString
  }

  private static let excludedRemoteMessagePrefixes = [
    "approval:",
    "remote-approval:",
    "agent-recovery:",
    "stale-connector:",
  ]
}

final class AgentReplySpeechController {
  private struct Session {
    var target: AgentReplySpeechTarget
    var playbackSessionId = ""
    var enabled = false
    var observedText: String
    var inputClosed = false
  }

  private var active: Session?
  private var playbackSequence: UInt64 = 0

  func observe(_ target: AgentReplySpeechTarget?) -> AgentReplySpeechCommand {
    guard let target else {
      let previous = active
      active = nil
      return AgentReplySpeechCommand(
        cancelSessionId: previous.flatMap { $0.enabled ? $0.playbackSessionId : nil } ?? "",
        changedEntryIds: changedEntryIds(previous?.target.entryId)
      )
    }
    guard var session = active, session.target.responseId == target.responseId else {
      let previous = active
      active = Session(
        target: target,
        observedText: target.text
      )
      return AgentReplySpeechCommand(
        cancelSessionId: previous.flatMap { $0.enabled ? $0.playbackSessionId : nil } ?? "",
        changedEntryIds: changedEntryIds(previous?.target.entryId, target.entryId)
      )
    }

    let oldEntryId = session.target.entryId
    let wasComplete = session.target.complete
    let delta = appendedText(previous: session.observedText, current: target.text)
    session.target = target
    session.observedText = target.text
    let shouldAppend = session.enabled && !session.inputClosed && !delta.isEmpty
    let shouldFinish = session.enabled && target.complete && !wasComplete && !session.inputClosed
    if shouldFinish { session.inputClosed = true }
    active = session
    return AgentReplySpeechCommand(
      appendedText: shouldAppend ? delta : "",
      finishSessionId: shouldFinish ? session.playbackSessionId : "",
      scheduleCommitSessionId: shouldAppend && !target.complete ? session.playbackSessionId : "",
      changedEntryIds: oldEntryId == target.entryId ? [] : [oldEntryId, target.entryId]
    )
  }

  func toggle(_ target: AgentReplySpeechTarget) -> AgentReplySpeechCommand {
    var session = session(for: target)
    if session.enabled {
      session.enabled = false
      session.inputClosed = false
      active = session
      return AgentReplySpeechCommand(
        cancelSessionId: session.playbackSessionId,
        changedEntryIds: [session.target.entryId]
      )
    }
    return begin(session: session, text: target.text, complete: target.complete)
  }

  func readParagraph(_ target: AgentReplySpeechTarget, paragraph: String) -> AgentReplySpeechCommand {
    let speech = paragraph.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !speech.isEmpty else { return AgentReplySpeechCommand() }
    var session = session(for: target)
    session.observedText = target.text
    return begin(session: session, text: speech, complete: target.complete)
  }

  func disable(sessionId: String) -> Set<String> {
    guard var session = active, session.playbackSessionId == sessionId else { return [] }
    session.enabled = false
    session.inputClosed = false
    active = session
    return [session.target.entryId]
  }

  func isEnabled(_ target: AgentReplySpeechTarget) -> Bool {
    active?.target.responseId == target.responseId && active?.enabled == true
  }

  func isActive(_ target: AgentReplySpeechTarget) -> Bool {
    active?.target.responseId == target.responseId
  }

  private func session(for target: AgentReplySpeechTarget) -> Session {
    if var current = active, current.target.responseId == target.responseId {
      current.target = target
      current.observedText = target.text
      return current
    }
    return Session(
      target: target,
      observedText: target.text
    )
  }

  private func begin(
    session initialSession: Session,
    text: String,
    complete: Bool
  ) -> AgentReplySpeechCommand {
    var session = initialSession
    let previousPlaybackSessionId = session.enabled ? session.playbackSessionId : ""
    playbackSequence &+= 1
    session.playbackSessionId = Self.playbackSessionId(
      session.target.responseId,
      sequence: playbackSequence
    )
    session.enabled = true
    session.inputClosed = complete
    active = session
    return AgentReplySpeechCommand(
      cancelSessionId: previousPlaybackSessionId,
      beginSessionId: session.playbackSessionId,
      appendedText: text,
      finishSessionId: complete ? session.playbackSessionId : "",
      scheduleCommitSessionId: complete ? "" : session.playbackSessionId,
      changedEntryIds: [session.target.entryId]
    )
  }

  private func appendedText(previous: String, current: String) -> String {
    guard current != previous, current.hasPrefix(previous) else { return "" }
    return String(current.dropFirst(previous.count))
  }

  private func changedEntryIds(_ values: String?...) -> Set<String> {
    Set(values.compactMap { $0 }.filter { !$0.isEmpty })
  }

  private static func playbackSessionId(_ responseId: String, sequence: UInt64) -> String {
    let digest = SHA256.hash(data: Data(responseId.utf8))
    let suffix = digest.prefix(8).map { String(format: "%02x", $0) }.joined()
    return "agent-reply-\(suffix)-\(sequence)"
  }
}
